aten/src/ATen/cuda/CUDAFuture.h

```diff
@@ -0,0 +1,133 @@
+    // To avoid races with other threads that may be using the extractor, we
+    // won't modify it after it's first set.
+    if (dataPtrExtractor_ == nullptr) {
+      dataPtrExtractor_ = std::move(data_ptr_extractor);
+    }
+  }
+
+ protected:
+  c10::intrusive_ptr<Future> createInstance(at::TypePtr type) override {
+    auto fut = c10::make_intrusive<CUDAFuture>(std::move(type));
+    // The new future needs the DataPtr extractor when it gets marked complete
+    // but this might happen immediately inline or in parallel by another
+    // thread. In both these cases this would/might happen before the user has
+    // time to set their own DataPtr extractor, which might lead to failures
+    // if the default extractor can't handle some of the user's types.
+    // Therefore we propagate our extractor.
+    fut->setDataPtrExtractor(dataPtrExtractor_);
+    return fut;
+  }
+
+  void postMarkCompletedHook(const at::IValue& value) override {
+    std::vector<bool> isCudaDeviceUsed(c10::cuda::device_count(), false);
+    for (const at::DataPtr& data_ptr : extractDataPtrs(value)) {
+      if (data_ptr.device().is_cuda()) {
+        isCudaDeviceUsed[data_ptr.device().index()] = true;
+      }
+    }
+
+    cudaEvents_ = std::make_shared<std::vector<at::cuda::CUDAEvent>>();
+    for (c10::DeviceIndex idx = 0; idx < isCudaDeviceUsed.size(); idx++) {
+      if (isCudaDeviceUsed[idx]) {
+        at::cuda::CUDAEvent cudaEvent;
+        cudaEvent.record(at::cuda::getCurrentCUDAStream(idx));
+        (*cudaEvents_).push_back(std::move(cudaEvent));
+      }
+    }
+  }
+
+  std::function<void(void)> wrapCallback(
+      std::function<void(void)> callback) override {
+    return [this, callback{std::move(callback)}]() {
+      // Get a stream for all devices, even those that are not used by the
+      // value, because the user's callback could use those other devices.
+      std::vector<at::cuda::CUDAStream> streams;
+      for (c10::DeviceIndex idx = 0; idx < c10::cuda::device_count(); idx++) {
+        // FIXME Should we find a way to allow to change the priority of
+        // streams?
+        streams.push_back(
+            at::cuda::getStreamFromPool(/*isHighPriority=*/false, idx));
+      }
+
+      // Do not free the underlying data storage of value_ before its
+      // usage on the stream finishes.
+      for (const at::DataPtr& data_ptr : extractDataPtrs(constValue())) {
+        if (data_ptr.device().is_cuda()) {
+          c10::cuda::CUDACachingAllocator::recordStream(
+              data_ptr, streams[data_ptr.device().index()]);
+        }
+      }
+
+      for (at::cuda::CUDAEvent& cudaEvent : *cudaEvents_) {
+        cudaEvent.block(streams[cudaEvent.device_index()]);
+      }
+
+      // Use the dedicated callback stream to run callback.
+      at::cuda::CUDAMultiStreamGuard streamGuard(streams);
+
+      callback();
+    };
+  }
+
+  void postWaitHook() override {
+    for (at::cuda::CUDAEvent& cudaEvent : *cudaEvents_) {
```

Comment:
hmm, this means users cannot call wait on `CUDAFuture` before it is marked as completed?

Reply:
Not sure I follow: the `postWaitHook` is called by ivalue::Future once a wait has finished, and that only happens when the future is marked complete. Thus one can certainly start waiting before the future is complete, in which case one will block inside ivalue::Future (waiting on a condition variable). Since that only unblocks after the future is complete, we're guaranteed that `postWaitHook` will be called _after_ `postMarkCompletedHook`. Is this what you were wondering about?